rollingpin/deploy.py

```diff
@@ -118,21 +119,50 @@ def signal_handler(sig, _):
                 yield self.event_bus.trigger("build.begin")
 
                 try:
-                    build_command = ["build"] + components
-                    (tokens,) = yield self.process_host(
-                        self.build_host, [build_command])
-
+                    # synchronize the code host with upstreams
+                    # this will return a build token and build host for each
+                    # component
+                    sync_command = ["synchronize"] + components
+                    (sync,) = yield self.process_host(
+                        self.code_host, [sync_command])
+
+                    # this is where we build up the final deploy command
+                    # resulting from all our syncing and building
                     deploy_command = ["deploy"]
-                    for component in components:
-                        try:
-                            deploy_ref = component + "@" + tokens[component]
-                        except KeyError:
-                            raise ComponentNotBuiltError(component)
-                        deploy_command.append(deploy_ref)
-                    commands = [deploy_command] + commands
+
+                    # collect the results of the sync per-buildhost
+                    by_buildhost = collections.defaultdict(list)
+                    for component, sync_info in sync.iteritems():
+                        component_ref = component + "@" + sync_info["token"]
+
+                        build_host = sync_info.get("buildhost", None)
+                        if build_host:
+                            by_buildhost[build_host].append(component_ref)
+                        else:
+                            # no build host means we just pass the build token
+                            # straight through as a deploy token
+                            deploy_command.append(component_ref)
+
+                    # ask each build host to build our components and return
+                    # a deploy token
+                    for build_host, build_refs in by_buildhost.iteritems():
+                        build_command = ["build"] + build_refs
+                        (tokens,) = yield self.process_host(
+                            build_host, [build_command])
+
+                        for ref in build_refs:
+                            try:
+                                deploy_ref = component + "@" + tokens[ref]
```

Comment:
crap, you're totally right. i should've tested with multiple components in the same build.